### PR TITLE
Add support for seastar::lw_shared_ptr<T>

### LINKED ISSUE #14: Expose seastar::lw_shared_ptr to Rust
`seastar::lw_shared_ptr<>` is a smart pointer that is used everywhere in seastar and scylla.
It would be great to add support for it, similarly to how `cxx` already supports `std::unique_ptr`.

It might require us to modify `cxx` as it doesn't support templates and has custom code in order to support `std::unique_ptr`.

### PR DESCRIPTION
**Pull request by sylwiaszunejko**
Fixes: [seastar-rs#14](https://github.com/zpp-2022-rust-seastar/seastar-rs/issues/14)

This PR introduces the Rust binding for `seastar::lw_shared_ptr<T>` that is called `SeastarLwSharedPtr<T>`.
It is analogous to the binding for `std::shared_ptr<T>`.
SeastarLwSharedPtr<T> does not support T being an opaque Rust type. You should use a Box<T> (C++ [rust::Box<T>]) instead for transferring ownership of opaque Rust types on the language boundary.

Example of usage:
```rust
// src/main.rs

use std::ops::Deref;
use std::ptr;

#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("example/src/example.hh");

        type Object;

        fn create_lw_shared_ptr() -> SeastarLwSharedPtr<Object>;
    }
}

#[test]
fn test_example() {
    let ptr1 = ffi::create_lw_shared_ptr();

    {
        // Create a second seastar::lw_shared_ptr holding shared ownership of the same
        // object. There is still only one Object but two SeastarLwSharedPtr<Object>.
        // Both pointers point to the same object on the heap.
        let ptr2 = ptr1.clone();
        assert!(ptr::eq(ptr1.deref(), ptr2.deref()));

        // ptr2 goes out of scope, but Object is not destroyed yet.
    }

    println!("say goodbye to Object");

    // ptr1 goes out of scope and Object is destroyed.
}

```
```c++
// include/example.hh

#pragma once
#include <memory>
#include <seastar/core/shared_ptr.hh>

class Object {
public:
Object();
~Object();
};

seastar::lw_shared_ptr<Object> create_lw_shared_ptr();
```
```c++
// src/example.cc

#include "example.hh"
#include <iostream>

Object::Object() { std::cout << "construct Object" << std::endl; }
Object::~Object() { std::cout << "~Object" << std::endl; }

seastar::lw_shared_ptr<Object> create_lw_shared_ptr() {
  return seastar::make_lw_shared<Object>();
}

```

### CODE REVIEW

File: src/seastar_lw_shared_ptr.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+use core::mem::MaybeUninit;
+use core::ops::Deref;
+
+/// Binding to C++ `seastar::lw_shared_ptr<T>`.
+#[repr(C)]
+pub struct SeastarLwSharedPtr<T>
+where
+    T: SeastarLwSharedPtrTarget,
+{
+    repr: [MaybeUninit<*mut c_void>; 2],
```
Comment (Lorak-mmk) [inline src/seastar_lw_shared_ptr.rs]: I assume that this array is here so that the struct has correct size. Why does it have 2 elements then? You put the following assert in `gen/src/write.rs`:
```
"static_assert(sizeof(::seastar::lw_shared_ptr<{}>) == sizeof(void *), \"\");",
```
Version for `std::shared_ptr` has `2 * sizeof(void *)` - so I assume that this array should also have just 1 element?

File: tests/ffi/tests.cc [hunk 2]
```diff
@@ -872,7 +876,7 @@ extern "C" const char *cxx_run_test() noexcept {
   ASSERT(cstr == "foo");
   ASSERT(other_cstr == "test");
 
-  const char *utf8_literal = u8"Test string";
+  const char *utf8_literal = reinterpret_cast<const char*>(u8"Test string");
```
Comment (Lorak-mmk) [inline tests/ffi/tests.cc:879]: Why is this change required?

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]

File: tests/ffi/tests.cc (hunk 2 above)
Comment (sylwiaszunejko) [reply tests/ffi/tests.cc:879]: Because we need to use C++20 to compile Seastar, and this type of cast is not allowed in C++20

File: src/seastar_lw_shared_ptr.rs (hunk 1 above)
Comment (sylwiaszunejko) [reply src/seastar_lw_shared_ptr.rs]: Yes, that was mistake, I fixed it

**Review by Lorak-mmk — APPROVED** [review]
